# NovaTel Customer Retention & Revenue Intelligence System

## Exploratory Data Analysis

### Objective

The objective of this notebook is to identify patterns, trends, customer behavior, churn drivers, loyalty indicators, product performance, and revenue opportunities.

### Analysis Areas

- Customer Analysis
- Churn Analysis
- Revenue Analysis
- Loyalty Analysis
- Product Analysis
- Geographic Analysis

In [2]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option("display.max_columns", None)

In [3]:
df = pd.read_csv("../data/raw/telecom_customer_churn.csv")

df.shape

(7043, 38)

## 1. Customer Status Analysis

### Purpose

Understand the distribution of customers across the customer lifecycle.

Questions:

- How many customers stayed?
- How many customers churned?
- How many customers joined?
- What percentage does each group represent?

In [4]:
status_counts = df["Customer Status"].value_counts()

status_pct = round(
    df["Customer Status"].value_counts(normalize=True) * 100,
    2
)

pd.DataFrame({
    "Count": status_counts,
    "Percentage": status_pct
})

,Count,Percentage
Customer Status,,
Stayed,4720,67.02
Churned,1869,26.54
Joined,454,6.45


### Findings

- 67.02% of customers remained with NovaTel.
- 26.54% of customers churned.
- 6.45% of customers recently joined.

### Business Insight

- The retained customer base remains the largest segment.
- Customer churn represents a significant business challenge.
- More than one-quarter of customers have left the company.
- Understanding churn drivers should be a primary focus of the analysis.

## 2. Revenue by Customer Status

### Purpose

Understand how revenue is distributed across customer lifecycle stages.

Questions:

- How much revenue comes from retained customers?
- How much revenue has been lost through churn?
- How much revenue is associated with new customers?

In [5]:
revenue_by_status = (
    df.groupby("Customer Status")["Total Revenue"]
    .sum()
    .sort_values(ascending=False)
    .round(2)
)

revenue_by_status

Customer Status
Stayed     17632392.12
Churned     3684459.82
Joined        54279.75
Name: Total Revenue, dtype: float64

### Findings

- Retained customers generated 17.63 million in total revenue.
- Churned customers generated 3.68 million in total revenue.
- Joined customers generated 54.28 thousand in total revenue.

### Business Insight

- Customer retention is the primary driver of company revenue.
- Churn has resulted in a substantial loss of revenue.
- Revenue protection should be a key objective of NovaTel's retention strategy.

### Revenue Contribution Analysis

Determine the percentage contribution of each customer status segment to total revenue.

In [6]:
revenue_pct = round(
    df.groupby("Customer Status")["Total Revenue"]
      .sum()
      / df["Total Revenue"].sum()
      * 100,
    2
)

revenue_pct

Customer Status
Churned    17.24
Joined      0.25
Stayed     82.51
Name: Total Revenue, dtype: float64

### Findings

- Retained customers contribute 82.51% of total revenue.
- Churned customers account for 17.24% of total revenue.
- Joined customers account for 0.25% of total revenue.

### Business Insight

- NovaTel's revenue is heavily dependent on customer retention.
- A significant portion of revenue is associated with customers who have already left the company.
- Improving retention could protect a substantial share of company revenue.

## 3. Churn Reason Analysis

### Purpose

Identify the primary reasons customers leave NovaTel.

Questions:

- What are the most common churn reasons?
- Which churn categories drive the largest losses?
- Where should retention efforts be focused?

In [7]:
df["Churn Category"].value_counts(dropna=False)

Churn Category
NaN                5174
Competitor          841
Dissatisfaction     321
Attitude            314
Price               211
Other               182
Name: count, dtype: int64

### Findings

- Competitor-related churn is the largest churn category with 841 customers.
- Competitor churn accounts for approximately 45% of all churned customers.
- Dissatisfaction and Attitude are the second and third largest churn categories.
- Price-related churn represents a smaller portion of total churn.

### Business Insight

- Competitive pressure is the primary driver of customer loss.
- Retention initiatives should prioritize competitive positioning, offers, pricing strategy, and product differentiation.
- Reducing competitor-driven churn would likely produce the largest retention impact.

### Detailed Churn Reason Analysis

Identify the specific reasons customers left NovaTel.

In [8]:
df["Churn Reason"].value_counts().head(10)

Churn Reason
Competitor had better devices                313
Competitor made better offer                 311
Attitude of support person                   220
Don't know                                   130
Competitor offered more data                 117
Competitor offered higher download speeds    100
Attitude of service provider                  94
Price too high                                78
Product dissatisfaction                       77
Network reliability                           72
Name: count, dtype: int64

### Findings

Top churn reasons include:

1. Competitor had better devices (313)
2. Competitor made better offer (311)
3. Attitude of support person (220)
4. Don't know (130)
5. Competitor offered more data (117)

### Business Insight

- Competitor-related reasons dominate customer churn.
- Device offerings and promotional offers are the largest drivers of customer loss.
- Service experience also contributes to churn through support and provider attitude issues.
- NovaTel's retention strategy should focus on competitive differentiation and customer experience improvements.

## 4. Contract Analysis

### Purpose

Determine how contract type influences customer retention and churn behavior.

Questions:

- Which contract type experiences the highest churn?
- Which contract type retains customers most effectively?
- How does customer status vary across contract types?

In [9]:
pd.crosstab(
    df["Contract"],
    df["Customer Status"]
)

Customer Status,Churned,Joined,Stayed
Contract,,,
Month-to-Month,1655,408,1547
One Year,166,24,1360
Two Year,48,22,1813


### Findings

- Month-to-Month contracts account for 1,655 churned customers.
- Approximately 88.55% of all churn originates from Month-to-Month customers.
- One Year contracts experience significantly lower churn.
- Two Year contracts show the strongest customer retention performance.

### Business Insight

- Contract commitment is one of the strongest predictors of customer retention.
- Month-to-Month customers represent NovaTel's highest-risk segment.
- Increasing migration from Month-to-Month contracts to longer-term contracts could significantly reduce churn.

### Contract Churn Rate Analysis

Calculate churn rate for each contract type.

In [10]:
contract_churn_rate = round(
    (
        pd.crosstab(df["Contract"], df["Customer Status"])["Churned"]
        /
        pd.crosstab(df["Contract"], df["Customer Status"]).sum(axis=1)
    ) * 100,
    2
)

contract_churn_rate

Contract
Month-to-Month    45.84
One Year          10.71
Two Year           2.55
dtype: float64

### Findings

Churn Rate by Contract:

| Contract Type | Churn Rate |
|---------------|------------|
| Month-to-Month | 45.84% |
| One Year | 10.71% |
| Two Year | 2.55% |

### Business Insight

- Month-to-Month customers experience the highest churn rate.
- Nearly half of Month-to-Month customers eventually leave NovaTel.
- One-Year contracts substantially improve retention.
- Two-Year contracts provide the strongest customer retention performance.
- Contract commitment appears to be one of the strongest predictors of churn.

## 5. Tenure Analysis

### Purpose

Determine how customer tenure influences retention and churn behavior.

Questions:

- Do newer customers churn more often?
- Do long-term customers remain more loyal?
- How does tenure differ by customer status?

In [11]:
df.groupby("Customer Status")["Tenure in Months"] \
  .agg(["count", "mean", "median", "min", "max"]) \
  .round(2)

,count,mean,median,min,max
Customer Status,,,,,
Churned,1869,17.98,10.0,1,72
Joined,454,1.72,1.0,1,3
Stayed,4720,41.04,42.0,4,72


### Findings

- Churned customers have an average tenure of 17.98 months.
- The median tenure of churned customers is only 10 months.
- Retained customers have an average tenure of 41.04 months.
- The median tenure of retained customers is 42 months.

### Business Insight

- Customer churn is concentrated among newer customers.
- Long-tenure customers demonstrate significantly higher loyalty.
- The first year of the customer lifecycle appears to be the highest-risk retention period.
- Early retention initiatives could significantly reduce churn.

## 6. Referral Analysis

### Purpose

Determine whether customer referrals are associated with customer loyalty and retention.

Questions:

- Do highly referred customers churn less?
- Are referrals a signal of customer satisfaction?
- Can referral programs improve retention?

In [12]:
df.groupby("Customer Status")["Number of Referrals"] \
  .agg(["count", "mean", "median", "max"]) \
  .round(2)

,count,mean,median,max
Customer Status,,,,
Churned,1869,0.52,0.0,9
Joined,454,0.95,0.0,10
Stayed,4720,2.61,1.0,11


### Findings

- Retained customers average 2.61 referrals.
- Churned customers average only 0.52 referrals.
- Joined customers average 0.95 referrals.
- Retained customers generate substantially more referrals than churned customers.

### Business Insight

- Referrals appear to be a strong indicator of customer loyalty.
- Customers who refer others are significantly more likely to remain with NovaTel.
- Referral programs may improve both customer acquisition and retention.
- Referral behavior can potentially be used as an early indicator of customer satisfaction.

## 7. Internet Type Analysis

### Purpose

Determine how internet technology influences customer retention and churn.

Questions:

- Which internet type experiences the highest churn?
- Which internet type retains customers most effectively?
- Are specific products associated with churn risk?

In [13]:
pd.crosstab(
    df["Internet Type"],
    df["Customer Status"]
)

Customer Status,Churned,Joined,Stayed
Internet Type,,,
Cable,213,56,561
DSL,307,115,1230
Fiber Optic,1236,101,1698


### Findings

- Fiber Optic customers account for 1,236 churned customers.
- Fiber Optic represents the largest source of customer churn.
- DSL and Cable customers contribute substantially fewer churned customers.

### Preliminary Business Insight

- Fiber Optic customers represent a high-risk customer segment.
- Churn behavior should be further evaluated using churn rates rather than customer counts alone.

### Internet Type Churn Rate Analysis

Calculate churn rate by internet technology.

In [14]:
internet_churn_rate = round(
    (
        pd.crosstab(df["Internet Type"], df["Customer Status"])["Churned"]
        /
        pd.crosstab(df["Internet Type"], df["Customer Status"]).sum(axis=1)
    ) * 100,
    2
)

internet_churn_rate

Internet Type
Cable          25.66
DSL            18.58
Fiber Optic    40.72
dtype: float64

### Findings

Churn Rate by Internet Type:

| Internet Type | Churn Rate |
|--------------|------------|
| Fiber Optic | 40.72% |
| Cable | 25.66% |
| DSL | 18.58% |

### Business Insight

- Fiber Optic customers experience the highest churn rate.
- Fiber Optic churn is more than double DSL churn.
- Internet product strategy appears to be strongly linked to customer retention.
- Product competitiveness and service value should be reviewed for Fiber Optic customers.

## 8. Revenue at Risk Analysis

### Purpose

Measure the financial impact of customer churn.

Questions:

- How much revenue is associated with churned customers?
- How does revenue differ by customer status?
- What is the financial cost of customer attrition?

In [15]:
df.groupby("Customer Status")["Total Revenue"] \
  .agg(["count", "sum", "mean", "median"]) \
  .round(2)

,count,sum,mean,median
Customer Status,,,,
Churned,1869,3684459.82,1971.35,893.55
Joined,454,54279.75,119.56,91.10
Stayed,4720,17632392.12,3735.68,2960.11


### Findings

- Churned customers generated 3.68 million in total revenue.
- Retained customers generated 17.63 million in total revenue.
- Retained customers produce substantially higher average revenue than churned customers.
- Churned customers appear to leave before reaching their full lifetime value.

### Business Insight

- Customer churn creates significant revenue loss.
- Revenue growth depends heavily on retaining existing customers.
- Increasing retention could improve both customer lifetime value and total revenue.
- Retention initiatives may provide higher returns than customer acquisition alone.

## 9. Offer Analysis

### Purpose

Evaluate the effectiveness of marketing offers on customer retention and churn.

### Business Questions

- Which offers are associated with the lowest churn?
- Which offers retain customers most effectively?
- Do specific offers influence customer loyalty?
- Which offers should NovaTel continue, improve, or discontinue?

### Analysis Approach

Compare customer status distribution across offer types to identify potential relationships between promotional offers and retention outcomes.

In [16]:
pd.crosstab(
    df["Offer"],
    df["Customer Status"]
)

Customer Status,Churned,Joined,Stayed
Offer,,,
Offer A,35,0,485
Offer B,101,0,723
Offer C,95,0,320
Offer D,161,0,441
Offer E,426,175,204


### Findings

- Offer A and Offer B show strong retention patterns.
- Offer E contains substantially more churned customers than retained customers.
- Customer outcomes vary significantly across offer types.
- Marketing campaigns appear to influence retention behavior.

### Preliminary Business Insight

- Offer performance is not uniform across the customer base.
- Certain offers appear more effective at retaining customers.
- Offer E may require further investigation due to its high churn volume.

### Offer Churn Rate Analysis

Calculate churn rate for each offer type to measure retention effectiveness.

In [17]:
offer_churn_rate = round(
    (
        pd.crosstab(df["Offer"], df["Customer Status"])["Churned"]
        /
        pd.crosstab(df["Offer"], df["Customer Status"]).sum(axis=1)
    ) * 100,
    2
)

offer_churn_rate.sort_values(ascending=False)

Offer
Offer E    52.92
Offer D    26.74
Offer C    22.89
Offer B    12.26
Offer A     6.73
dtype: float64

### Findings

Churn Rate by Offer:

| Offer | Churn Rate |
|--------|------------|
| Offer A | 6.73% |
| Offer B | 12.26% |
| Offer C | 22.89% |
| Offer D | 26.74% |
| Offer E | 52.92% |

### Business Insight

- Offer A demonstrates the strongest retention performance.
- Offer B also performs well and is associated with low churn.
- Offer E has the highest churn rate and appears ineffective at retaining customers.
- Marketing offer strategy has a significant impact on customer retention outcomes.
- Offer optimization represents a major opportunity for churn reduction.

## 10. Payment Method Analysis

### Purpose

Determine whether payment methods influence customer churn behavior.

### Business Questions

- Which payment method has the highest churn?
- Which payment method retains customers best?
- Are payment preferences associated with customer loyalty?

In [18]:
pd.crosstab(
    df["Payment Method"],
    df["Customer Status"]
)

Customer Status,Churned,Joined,Stayed
Payment Method,,,
Bank Withdrawal,1329,181,2399
Credit Card,398,231,2120
Mailed Check,142,42,201


In [19]:
payment_churn_rate = round(
    (
        pd.crosstab(df["Payment Method"], df["Customer Status"])["Churned"]
        /
        pd.crosstab(df["Payment Method"], df["Customer Status"]).sum(axis=1)
    ) * 100,
    2
)

payment_churn_rate.sort_values(ascending=False)

Payment Method
Mailed Check       36.88
Bank Withdrawal    34.00
Credit Card        14.48
dtype: float64

### Findings

Churn Rate by Payment Method:

| Payment Method | Churn Rate |
|----------------|------------|
| Mailed Check | 36.88% |
| Bank Withdrawal | 34.00% |
| Credit Card | 14.48% |

### Business Insight

- Credit Card customers exhibit the strongest retention performance.
- Mailed Check customers experience the highest churn rate.
- Payment behavior appears to be associated with customer retention.
- Encouraging digital payment adoption may improve customer retention.

## 11. Age Group Analysis

### Purpose

Determine how customer churn varies across age groups.

### Business Questions

- Which age groups experience the highest churn?
- Are younger or older customers more likely to leave?
- Should retention strategies be targeted by age segment?

In [20]:
df["Age Group"] = pd.cut(
    df["Age"],
    bins=[18,30,45,60,80],
    labels=["19-30","31-45","46-60","61-80"]
)

pd.crosstab(
    df["Age Group"],
    df["Customer Status"]
)

Customer Status,Churned,Joined,Stayed
Age Group,,,
19-30,342,105,1082
31-45,457,161,1325
46-60,464,125,1320
61-80,606,63,993


In [21]:
age_churn_rate = round(
    (
        pd.crosstab(df["Age Group"], df["Customer Status"])["Churned"]
        /
        pd.crosstab(df["Age Group"], df["Customer Status"]).sum(axis=1)
    ) * 100,
    2
)

age_churn_rate.sort_values(ascending=False)

Age Group
61-80    36.46
46-60    24.31
31-45    23.52
19-30    22.37
dtype: float64

### Findings

Churn Rate by Age Group:

| Age Group | Churn Rate |
|------------|------------|
| 61-80 | 36.46% |
| 46-60 | 24.31% |
| 31-45 | 23.52% |
| 19-30 | 22.37% |

### Business Insight

- Customers aged 61-80 experience the highest churn rate.
- Senior customers appear substantially more likely to leave NovaTel than younger customers.
- Age appears to be an important demographic factor influencing retention.
- Retention initiatives may benefit from age-specific customer engagement strategies.

## 12. Unlimited Data Analysis

### Purpose

Determine whether unlimited data plans influence customer retention and churn behavior.

### Business Questions

- Do customers with unlimited data churn less?
- Does unlimited data improve customer loyalty?
- Could plan design be contributing to churn?

In [22]:
pd.crosstab(
    df["Unlimited Data"],
    df["Customer Status"]
)

Customer Status,Churned,Joined,Stayed
Unlimited Data,,,
No,254,48,470
Yes,1502,224,3019


In [23]:
unlimited_data_churn_rate = round(
    (
        pd.crosstab(df["Unlimited Data"], df["Customer Status"])["Churned"]
        /
        pd.crosstab(df["Unlimited Data"], df["Customer Status"]).sum(axis=1)
    ) * 100,
    2
)

unlimited_data_churn_rate.sort_values(ascending=False)

Unlimited Data
No     32.90
Yes    31.65
dtype: float64

### Findings

Churn Rate by Unlimited Data:

| Unlimited Data | Churn Rate |
|---------------|------------|
| No | 32.90% |
| Yes | 31.65% |

### Business Insight

- Unlimited Data shows minimal impact on churn behavior.
- Churn rates are similar across both groups.
- Unlimited Data does not appear to be a major driver of customer retention.
- Other factors such as contract type, offers, referrals, and internet type have a much stronger relationship with churn.

## 13. Geographic Analysis

### Purpose

Identify geographic areas contributing the most customer churn.

### Business Questions

- Which cities generate the most churn?
- Are specific geographic regions higher risk?
- Can retention efforts be geographically targeted?

In [24]:
df[df["Customer Status"] == "Churned"] \
    .groupby("City")["Customer ID"] \
    .count() \
    .sort_values(ascending=False) \
    .head(10)

City
San Diego        185
Los Angeles       78
San Francisco     31
San Jose          29
Fallbrook         26
Sacramento        26
Temecula          22
Escondido         16
Long Beach        15
Oakland           13
Name: Customer ID, dtype: int64

### Findings

Top Cities by Churn Volume:

| City | Churned Customers |
|---------|---------:|
| San Diego | 185 |
| Los Angeles | 78 |
| San Francisco | 31 |
| San Jose | 29 |
| Fallbrook | 26 |

### Business Insight

- San Diego contributes the largest number of churned customers.
- Large metropolitan areas account for a significant portion of total churn volume.
- Geographic targeting may improve the effectiveness of retention campaigns.
- Additional analysis would be required to determine city-level churn rates.